Sources/SwiftBasicFormat/BasicFormat.swift

```diff
@@ -0,0 +1,278 @@
+      return false
+    default:
+      break
+    }
+
+    return token.requiresTrailingSpace
+  }
+
+  // MARK: - Formatting a token
+
+  open override func visit(_ token: TokenSyntax) -> TokenSyntax {
+    lazy var previousTokenWillEndWithBlank: Bool = {
+      guard let previousToken = token.previousToken(viewMode: .sourceAccurate) else {
+        return false
+      }
+      return previousToken.trailingTrivia.pieces.last?.isBlank ?? false
+        || requiresTrailingBlank(previousToken)
+    }()
+
+    lazy var previousTokenWillEndInNewline: Bool = {
+      guard let previousToken = token.previousToken(viewMode: .sourceAccurate) else {
+        // Assume that the start of the tree is equivalent to a newline so we
+        // don't add a leading newline to the file.
+        return true
+      }
+      return previousToken.trailingTrivia.pieces.last?.isNewline ?? false
+    }()
+
+    lazy var nextTokenWillStartWithBlank: Bool = {
+      guard let nextToken = token.nextToken(viewMode: .sourceAccurate) else {
+        return false
+      }
+      return nextToken.leadingTrivia.first?.isBlank ?? false
+        || requiresLeadingBlank(nextToken)
+        || requiresLeadingNewline(nextToken)
+    }()
+
+    lazy var nextTokenWillStartWithNewline: Bool = {
+      guard let nextToken = token.nextToken(viewMode: .sourceAccurate) else {
+        return false
+      }
+      return nextToken.leadingTrivia.first?.isNewline ?? false
+        || requiresLeadingNewline(nextToken)
+    }()
+
+    lazy var trailingTriviaAndNextTokensLeadingTriviaWhitespace: Trivia = {
+      let nextToken = token.nextToken(viewMode: .sourceAccurate)
+      let nextTokenLeadingWhitespace = nextToken?.leadingTrivia.prefix(while: { $0.isIndentationWhitespace }) ?? []
+      return trailingTrivia + Trivia(pieces: nextTokenLeadingWhitespace)
+    }()
+
+    var leadingTrivia = token.leadingTrivia
+    var trailingTrivia = token.trailingTrivia
+
+    if requiresLeadingNewline(token) {
+      // Add a leading newline if the token requires it unless
+      //  - it already starts with a newline or
+      //  - the previous token ends with a newline
+      if !(leadingTrivia.first?.isNewline ?? false) && !previousTokenWillEndInNewline {
+        // Add a leading newline if the token requires it and
+        //  - it doesn't already start with a newline and
+        //  - the previous token didn't end with a newline
+        leadingTrivia = .newline + leadingTrivia
+      }
+    } else if requiresLeadingBlank(token) {
+      // Add a leading space if the token requires it unless
+      //  - it already starts with a blank or
+      //  - the previous token ends with a blank after the rewrite
+      if !(leadingTrivia.first?.isBlank ?? false) && !previousTokenWillEndWithBlank {
+        leadingTrivia += .space
+      }
+    }
+
+    if leadingTrivia.indentation(isOnNewline: previousTokenWillEndInNewline) == [] {
```

Comment:
If `containsIndentation` and `indentation` matched up, this could just be a contains check. It would be nice if we could remove anchor points, but I can't think of a fast way to do that (we can remove all tokens in the node in the post visit but... that doesn't seem worth the performance hit).